modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/MetaStorageManager.java

```diff
@@ -120,105 +142,237 @@
+        metaStorageNodeNamesFut.cancel(false);
 
-        // TODO: IGNITE-14414 Cluster initialization flow. Here we should complete metaStorageServiceFuture.
-//        clusterNetSvc.messagingService().addMessageHandler((message, senderAddr, correlationId) -> {});
+        try {
+            metaStorageSvcFut.get(10, TimeUnit.SECONDS);
+        }
+        catch (ExecutionException e) {
+            // execution exceptions are not important and are expected to be thrown, e.g. when the component stops
+            // abruptly
+            LOG.debug("Exception during MetaStorageManager stop", e.getCause());
+        }
+        catch (InterruptedException | TimeoutException e) {
+            throw new IgniteInternalException("Exception during MetaStorageManager stop", e);
+        }
     }
 
     /** {@inheritDoc} */
     @Override public void stop() {
-        // TODO: IGNITE-15161 Implement component's stop.
+        // no-op
+    }
+
+    /**
+     * Temporary method for emulating the "init" command.
+     *
+     * @param metastorageNodeNames Names of the nodes hosting the Meta Storage.
+     */
+    // TODO: remove this method: https://issues.apache.org/jira/browse/IGNITE-14414
+    public void init(List<String> metastorageNodeNames) {
+        assert metaStorageSvcFut != null : "MetaStorageManager has not been started";
+
+        metaStorageNodeNamesFut.complete(metastorageNodeNames);
+    }
+
+    /**
+     * Tries to repeatedly retrieve information about the nodes hosting the Meta Storage from a random cluster member.
+     */
+    private CompletableFuture<Collection<String>> joinCluster() {
+        if (metaStorageNodeNamesFut.isDone())
+            return metaStorageNodeNamesFut;
+
+        ClusterNode randomMember = randomMember();
+
+        // if a random member is null, then this is the first node in the cluster and there's nothing to be done
+        // until the "init" command arrives
+        if (randomMember == null)
+            return metaStorageNodeNamesFut;
+
+        MetastorageNodesRequest request = messagesFactory.metastorageNodesRequest().build();
+
+        LOG.info("Issuing a join request to {}", randomMember.address());
+
+        return clusterNetSvc.messagingService()
+            .invoke(randomMember, request, 1000)
+            .handle((response, ex) -> {
+                if (response instanceof MetastorageNodesResponse) {
+                    var metastorageResponse = (MetastorageNodesResponse) response;
+
+                    metaStorageNodeNamesFut.complete(metastorageResponse.metastorageNodes());
+
+                    return metaStorageNodeNamesFut;
+                }
+                else {
+                    if (ex != null)
+                        LOG.debug("Exception while joining the cluster", ex);
+                    else
+                        LOG.debug("Received a {} response", response.getClass());
```

Comment:
Don't we check that debug level is enabled before logging?

Reply:
why do we need to that here? We usually do this check to avoid doing unnecessary stuff, I don't think there is any here